instructions/fault-injection/fault-injection.adoc

```diff
@@ -0,0 +1,150 @@
+customer => preference => recommendation v1 from '6976858b48-dswhz': 141
+customer => Error: 503 - preference => Error: 503 - fault filter abort
+customer => Error: 503 - preference => Error: 503 - fault filter abort
+customer => preference => recommendation v2 from '7778d6fb89-j6lts': 141
+----
+
+=== Kiali's Graph
+
+Within the Kiali UI select the *Graph* option from the left hand navigation and then choose
+
+* Namespace: istio-tutorial
+* Versioned app graph
+* Requests percentage
+* Last 1m
+* Every 10s
+
+[#img-503]
+.Kiali Graph Showing 503 Failures
+image::503.png[]
+
+Note the 50% failure rate from preference to recommendation
+
+=== Clean up
+
+[source,bash,subs="+macros,+attributes"]
+----
+$ oc -n istio-tutorial delete -f src/istiofiles/virtual-service-recommendation-503.yml
+virtualservice.networking.istio.io "recommendation" deleted
+----
+
+[#delay]
+== Delay
+
+The most insidious of possible distributed computing faults is not a "down" service but a service that is responding slowly, potentially causing a cascading failure in your network of services.
+
+[source,bash,subs="+macros,+attributes"]
+----
+$ oc -n istio-tutorial apply -f src/istiofiles/virtual-service-recommendation-delay.yml
+virtualservice.networking.istio.io/recommendation created
+----
+
+And hit the customer endpoint
+
+[source,bash,subs="+macros,+attributes"]
+----
+$ while :; do curl http://${INGRESS_GATEWAY} ; done
+----
+
+You will notice many requests to the customer endpoint now have a delay.
+If you are monitoring the logs for recommendation v1, v2 and v3, you will also see the delay happens BEFORE the recommendation service is actually called
+
+[source,bash,subs="+macros,+attributes"]
+----
+$ oc -n istio-tutorial logs -f recommendation-v#-XXXXXXXXXXXX -c recommendation
+Apr 26, 2019 9:31:09 PM com.redhat.developer.demos.recommendation.rest.RecommendationResource getRecommendations
+INFO: recommendation request from 7fbb8f794-ngw58: 8334
+Apr 26, 2019 9:31:17 PM com.redhat.developer.demos.recommendation.rest.RecommendationResource getRecommendations
+INFO: recommendation request from 7fbb8f794-ngw58: 8335
+----
+
+=== Kiali's Distributed Tracing
+
+Within the Kiali UI select the *Distributed Tracing* option from the left hand navigation and then choose
+
+* Namespace: istio-tutorial
+* Service: recommendation.istio-tutorial
+
+and finally press the *Search* button.
+
+[#img-503]
+.Kiali Graph Showing Delays
+image::delay.png[]
+
+Note that 50% of the traces are slightly over the artificial 7s delay while the other 50% are in the low ms range
+
+=== Clean up
+
+[source,bash,subs="+macros,+attributes"]
+----
+$ oc -n istio-tutorial delete -f src/istiofiles/destination-rule-recommendation.yml
+destinationrule.networking.istio.io "recommendation" deleted
+
+$ oc -n istio-tutorial delete -f src/istiofiles/virtual-service-recommendation-delay.yml
+virtualservice.networking.istio.io "recommendation" deleted
```

Comment:
Same comment as the 'oc create', can we combine them into a single yaml file?

Reply:
I apply/delete the destination rule once and use it for the apply/delete of the 2 virtual services yamls

Reply:
I know but it's still simpler to have one yaml file since each would then be a single pair of 'oc create -f fault-injection-503.yaml'/'oc create -f fault-injection-503.yaml'.

If we start mixing/matching multiple yaml files then someone will get it wrong.

Reply:
OK, I'll combine them and have the destination rule in both. Also, I think we should always use apply instead of create.

Reply:
Why apply rather than create?

Reply:
Maybe it's just personal preference, but if you do create/edit/create then you get an error that it already exists. If you do apply/edit/apply then it updates.

Reply:
I guess, my preference is for it to fail if there is a clash and I don't think we would be getting them to edit anything.  Either way we need to be consistent across the docs.